File: src/parser/helpers/nodes.js
"use strict";

/**
 * Create parsing node object.
 *
 * @param  {object} S -State object.
 * @param  {string} type - The object type to create.
 * @return {object} - The created node object.
 */
module.exports = (type, S) => {
	let N = {
		node: type,
		line: S.line,
		start: S.i,
		end: -1
	};

	/**
	 * Creates Node.value object.
	 *
	 * @return {object} The Node.value object.
	 */
	let o = () => ({ start: -1, end: -1, value: "" });

	// Modify each type as needed.
	switch (type) {
		case "COMMENT":
			N.comment = o();
			N.inline = false;

			break;

		case "NEWLINE":
			break;

		case "SETTING":
			N.sigil = o();
			N.name = o();
			N.assignment = o();
			N.value = o();
			N.args = [];

			break;

		case "VARIABLE":
			N.sigil = o();
			N.name = o();
			N.assignment = o();
			N.value = o();
			N.args = [];

			break;

		case "COMMAND":
			N.command = o();
			N.name = o();
			N.brackets = o();
			N.assignment = o();
			N.delimiter = o();
			N.value = o();
			N.flags = [];

			break;

		case "FLAG":
			N.hyphens = o();
			N.variable = o();
			N.name = o();
			N.alias = o();
			N.boolean = o();
			N.assignment = o();
			N.delimiter = o();
			N.multi = o();
			N.brackets = o();
			N.value = o();
			N.keyword = o();
			N.singleton = false;
			N.virtual = false;
			N.args = [];

			break;

		case "OPTION":
			N.bullet = o();
			N.value = o();
			N.args = [];

			break;

		case "BRACE":
			N.brace = o();

			break;
	}

	return N;
};

File: src/parser/nim/parser/helpers/nodes.nim
# Create parsing node object.
#
# @param  {string} type - The object type to create.
# @param  {object} S -State object.
# @return {object} - The created node object.
proc node*(nkType: NodeKind, S: State): Node =
    result = Node(kind: nkType, line: S.line, start: S.i, stop: -1) # new(result)

    # [https://github.com/nim-lang/Nim/issues/11395]
    # [https://forum.nim-lang.org/t/2799#17448]
    case (nkType):

    # Define each Node's props: [https://forum.nim-lang.org/t/4381]
    # [https://nim-lang.org/docs/manual.html#types-reference-and-pointer-types]

    of nkComment:
        result.comment = Branch()

    of nkNewline, nkEmpty: discard

    of nkSetting:
        result.sigil = Branch()
        result.name = Branch()
        result.assignment = Branch()
        result.value = Branch()
        result.args = @[]

    of nkVariable:
        result.sigil = Branch()
        result.name = Branch()
        result.assignment = Branch()
        result.value = Branch()
        result.args = @[]

    of nkCommand:
        result.command = Branch()
        result.name = Branch()
        result.brackets = Branch()
        result.assignment = Branch()
        result.delimiter = Branch()
        result.value = Branch()
        result.flags = @[]

    of nkFlag:
        result.hyphens = Branch()
        result.variable = Branch()
        result.name = Branch()
        result.alias = Branch()
        result.boolean = Branch()
        result.assignment = Branch()
        result.delimiter = Branch()
        result.multi = Branch()
        result.brackets = Branch()
        result.value = Branch()
        result.keyword = Branch()
        result.singleton = false
        result.virtual = false
        result.args = @[]

    of nkOption:
        result.bullet = Branch()
        result.value = Branch()
        result.args = @[]

    of nkBrace:
        result.brace = Branch()
